homeassistant/components/google_drive/backup.py

```diff
@@ -0,0 +1,147 @@
+"""Backup platform for the Google Drive integration."""
+
+from __future__ import annotations
+
+from collections.abc import AsyncIterator, Callable, Coroutine
+import logging
+from typing import Any
+
+from google_drive_api.exceptions import GoogleDriveApiError
+
+from homeassistant.components.backup import AgentBackup, BackupAgent, BackupAgentError
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import ChunkAsyncStreamIterator
+from homeassistant.util import slugify
+
+from . import DATA_BACKUP_AGENT_LISTENERS, GoogleDriveConfigEntry
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_get_backup_agents(
+    hass: HomeAssistant,
+    **kwargs: Any,
+) -> list[BackupAgent]:
+    """Return a list of backup agents."""
+    entries = hass.config_entries.async_loaded_entries(DOMAIN)
+    return [GoogleDriveBackupAgent(entry) for entry in entries]
+
+
+@callback
+def async_register_backup_agents_listener(
+    hass: HomeAssistant,
+    *,
+    listener: Callable[[], None],
+    **kwargs: Any,
+) -> Callable[[], None]:
+    """Register a listener to be called when agents are added or removed.
+
+    :return: A function to unregister the listener.
+    """
+    hass.data.setdefault(DATA_BACKUP_AGENT_LISTENERS, []).append(listener)
+
+    @callback
+    def remove_listener() -> None:
+        """Remove the listener."""
+        hass.data[DATA_BACKUP_AGENT_LISTENERS].remove(listener)
+        if not hass.data[DATA_BACKUP_AGENT_LISTENERS]:
+            del hass.data[DATA_BACKUP_AGENT_LISTENERS]
+
+    return remove_listener
+
+
+class GoogleDriveBackupAgent(BackupAgent):
+    """Google Drive backup agent."""
+
+    domain = DOMAIN
+
+    def __init__(self, config_entry: GoogleDriveConfigEntry) -> None:
+        """Initialize the cloud backup sync agent."""
+        super().__init__()
+        assert config_entry.unique_id
+        self.name = config_entry.title
+        self.unique_id = slugify(config_entry.unique_id)
+        self._client = config_entry.runtime_data
+
+    async def async_upload_backup(
+        self,
+        *,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        backup: AgentBackup,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup.
+
+        :param open_stream: A function returning an async iterator that yields bytes.
+        :param backup: Metadata about the backup that should be uploaded.
+        """
+        try:
+            await self._client.async_upload_backup(open_stream, backup)
+        except (GoogleDriveApiError, HomeAssistantError, TimeoutError) as err:
+            _LOGGER.error("Upload backup error: %s", err)
```

Comment:
Don't log errors when raising a backup exception. Include the message in the exception instead.

Reply:
#136916